interactive bots: Create Merels bot.

Privisus opened the pull request:
Link to GCI task: https://codein.withgoogle.com/dashboard/task-instances/5792612038475776/

roberthoenig commented on zulip_bots/zulip_bots/bots/merels/.gitignore:
```diff
@@ -0,0 +1,3 @@
+*.db
+
+.idea/
```
Add a newline here.

Privisus replied:
Added.

roberthoenig commented on zulip_bots/zulip_bots/bots/merels/merels.py:
```diff
@@ -0,0 +1,29 @@
+        if 'subject' in message:
+            topic = message['subject']
+
+        response = game.beat(content, topic, self.merels_db)
```
It seems like this logic will make all private conversations share the same game. You can make this work for private conversations by extracting the private conversation key from the `message` object.

roberthoenig commented on zulip_bots/zulip_bots/bots/merels/libraries/database.py:
```diff
@@ -0,0 +1,149 @@
+"""This module is used for managing, storing, and operating certain
```
Recently, we added a feature to Zulip that allows bots to store data on a Zulip server permanently. You can read about it in our bot documentation, and `virtual_fs` and `tictactoe` are two bots using such a storage system. Would it be feasible for you to refactor your PR to use the bot storage system instead of sqlite?

Privisus replied:
Unfortunately it's not really feasible for me to do that:

* [When I read performance considerations on this link](https://www.zulipchat.com/api/writing-bots). this means that I have to refactor every function that opens and closes the database connection, which I think is many
* Since it's still experimental, I didn't really want to touch on feature for now to maintain stability
* I store many columns on a record, so it's a little bit messier to work with if I switch to the feature as it only stores key and value 
* test_database.py will have to be rewritten somehow to *integrate* zulip storage system, which I don't know how to do

Privisus replied:
I decided to port it anyway. Should be fine now.

roberthoenig commented on zulip_bots/zulip_bots/bots/merels/libraries/game.py:
```diff
@@ -0,0 +1,211 @@
+    if "help" in message.lower():
+        return getHelp()
+
+    if "reset" in message.lower():
```
The reset command didn't work in our tests: Even though we had a game started, it printed "No game created yet".

Privisus replied:
Sorry, forgot to put `not` in `is None` below that. Just what happens when you write it at 3 AM. Should work fine now.

roberthoenig commented on zulip_bots/zulip_bots/bots/merels/libraries/game.py:
```diff
@@ -0,0 +1,211 @@
+    :return: Help message
+    """
+
+    return "Commands: create, reset, put (v,h), move (v,h) -> (v,h), take (" \
```
This could be a bit more detailed, e.g. add some explanations to the individual commands, a quick note on the game merels itself, etc.

Privisus replied:
Added more info for the commands.

Privisus commented:
Changes:
* Added a new line in .gitignore
* Added more information on command usage
* Fixed reset command
* Ported the entire merels from sqlite to storage system
* Increased code coverage for `merels` from around 70% to 97%

roberthoenig commented:
@Privisus this is  great! Merging. There are still some of minor code style quirks, but that's natural with a PR as big as this one. The strategy is then usually to merge everything, and do subsequent improvements as follow-ups.

roberthoenig commented:
One note is that it is generally useful to focus testing on the "real-message" bot tests, since that's what matters in the end.

Privisus commented:
Given the time nature of the current situation, I can only hope for the best so yeah, thank you!
